### ISSUE TITLE
about ScrollView on ios 13 . scroll bar will be center sometimes for ios 13 and can move to phone left

### ISSUE THREAD

**Issue by KarenQiao** [facebook/react-native#26610]
"react": "16.8.3",
"react-native": "0.59.9",
  Platforms: iOS 13.0, 

![image](https://user-images.githubusercontent.com/27947842/65738561-e0493380-e114-11e9-81dc-2f385362d855.png)

**Comment by KarenQiao** [comment]
System:
      OS: macOS 10.14.5
      CPU: (4) x64 Intel(R) Core(TM) m3-7Y32 CPU @ 1.10GHz
      Memory: 43.45 MB / 8.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.13.0 - /usr/local/bin/node
      Yarn: 1.16.0 - /usr/local/bin/yarn
      npm: 6.4.1 - /usr/local/bin/npm
    SDKs:
      iOS SDK:
        Platforms: iOS 13.0, DriverKit 19.0, macOS 10.15, tvOS 13.0, watchOS 6.0
      Android SDK:
        API Levels: 23, 26, 27, 28
        Build Tools: 27.0.3, 28.0.3
        System Images: android-26 | Google APIs Intel x86 Atom, android-28 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: 3.4 AI-183.6156.11.34.5522156
      Xcode: 11.0/11A420a - /usr/bin/xcodebuild
    npmPackages:
      react: 16.8.3 => 16.8.3 
      react-native: 0.59.9 => 0.59.9 
    npmGlobalPackages:
      react-native-cli: 2.0.1

**Comment by fahmidme** [comment]
Same issue here

**Comment by amsdamsgram** [comment]
Hey, I have the same issue with `FlatList` component.

**Comment by cernadasjuan** [comment]
Same issue here

**Comment by losh11** [comment]
I'm actually experiencing this with `react-native-webview` from `react-native-community`... didn't know it affected more than just the webview.

**Comment by edo1493** [comment]
I can see the same on react-native 0.60.5 with FlatList.

**Comment by fahmidme** [comment]
I don't know if this happened with anyone else or is that related but incase anyone experienced it: I saw my custom contact list for a second and then it changed to something like this with the scroll bar in the middle and then the thing disappeared.

**Comment by mobitar** [comment]
This appears to have something to do with SafeAreaView's `forceInset` property, at least in my case. Removing that property causes this issue to not appear anymore.

Possibly related: https://github.com/facebook/react-native/issues/16997

**Comment by sbycrosz** [comment]
We can reproduce this on 59.10 on a ScrollView with `contentContainerStyle={[{ paddingBottom: 60 ]}` prop. 

We're not sure why but removing the prop, or navigating to another page and back fix the scroll indicator somehow.

**Comment by sbycrosz** [comment]
Updates:
We found that adding `scrollIndicatorInsets={{ right: 1 }}` will force the scrollbar to be rendered correctly
Additionally, adding `scrollIndicatorInsets={{ right: 0 }}` does nothing. So we guess there's buggy code in iOS 13 that overrides contentInsets default since we didn't use SafeAreaView in our app.

**Comment by esdrasevt** [comment]
> This appears to have something to do with SafeAreaView's `forceInset` property, at least in my case. Removing that property causes this issue to not appear anymore.
> 
> Changing `<SafeAreaView forceInset={{ bottom: 'never'}}>` to `<SafeAreaView forceInset={{ bottom: 'never', vertical: 'never'}}>` also fixed the issue for me.
> 
> Possibly related: #16997

I'm not sure about the `forceInset` property, but it's definitely related fo SafeAreaView since I did some tests on different device sizes on the simulator, and the problem only occurs on devices that have the "notch" on the top of the screen, thus, devices that use the SafeAreaView.

**Comment by Elijen** [comment]
Same issue here. It's not possible to reproduce with 100% success, but so far we noticed this on iPhone Xs running iOS 13.0 and 13.1 on `expo@35.0.0` & `react-native@0.59.8`

**Comment by marlti7** [comment]
Same issue here

**Comment by 18612705334** [comment]
samae issue here with   ios 13.1.3  "react": "16.3.1",   "react-native": "0.55.3",

**Comment by seedalri** [comment]
> Hey, I have the same issue with `FlatList` component.

Same happened to me, but the issue was with ScrollView. Removed the <ScrollView> as we used FlatList and the problem was gone.

**Comment by rachelyao** [comment]
Same issue here with ios 13, react 16.9.0, react-native 0.61.2

**Comment by chrisheninger** [comment]
I initially dug into this thinking it was something I did incorrectly– after stripping away SafeAreaView, using different versions of react-native, digging into react-navigation, and a million other little tests– I narrowed it down to the following:

- Only occurs on iOS 13
- Only occurs when a FlatList/ScrollView is scrolling into an "inset" area

Fiddling with SafeAreaView props and everything else didn't help– but @sbycrosz's workaround of adding `scrollIndicatorInsets={{ right: 1 }}` to the FlatList worked perfectly. (thanks sbycrosz! 🙇)

**Comment by stephenheron** [comment]
We just ran into this issue as well. On a physical iPhone Xs (iOS 13.2.3) the issue does not appear but on the iPhone 11 Pro (iOS 13.3) using the simulator it does appear. We are updating the physical iPhone at the moment to iOS 13.3 to see if we can replicate the issue. 

@sbycrosz's workaround fixes it but a bit annoying having to update our ScrollViews.

**Comment by EdPutans** [comment]
I had the same issue when a View component with `style={{ flex:  1 }}` was the highest component in the component tree.
Replacing it with `ScrollView` and setting  `contentContainerStyle={{ flexGrow: 1 }}` seems to "fix" it.

**Comment by asakalou-exadel** [comment]
I am experiencing the same issue on iOS13 iPhone Xs .How did you fix it for react-native-webview?
Thanks!

**Comment by damianrodrigues** [comment]
Had the same problem for iPhone XR in Simulator (13.3) using `ScrollView`. Adding `scrollIndicatorInsets={{ right: 1 }}` put the scrollbar in its place.

**Comment by hramos** [comment]
Closing after several months without followup from the author. Please make sure to fill in the issue template if you decide to file a new issue.

**Comment by shirohana** [comment]
Is it any related issue tracking this? I don't think that it is a out of date problem :/ 

---
Device: iPhone 11 Pro Max with iOS 13.3/Simulator
Version: ReactNative@0.61.5, React@16.9.0

---
![untitled](https://user-images.githubusercontent.com/11359892/73992643-b4736f80-498a-11ea-956a-bb68b648d4ef.png)

**Comment by ManAnRuck** [comment]
@hramos i also have still the same issue 🤨

**Comment by jasperkuperus** [comment]
Also still seeing this issue. Using RN 0.61.5.

@hramos Although the author has not used the issue template correctly, the issue is very clear to me and the others in here. Please re-open this. Here's a full issue template filled out for you:

### Description:

The vertical scrollbar of a `FlatList` does not always render correctly to the right side of the screen. See all the screenshots in this issue.

For me. This happened even when I slimmed down my code into a `<FlatList />` with only `<Text />`. It only happened though when my list was longer than the height of the screen. The scrollbar then suddenly moves to the left. The longer the list, the further the scrollbar moves to the left.

Also another condition for me. Upon a fresh first render of my `<FlatList />`, this behaviour appeared. When navigating away, effectively unmounting this component, and back, the scrollbar is suddenly OK. Close and open the app: The scrollbar is at the wrong position again.

### React Native version:

```
info Fetching system and libraries information...
System:
    OS: macOS 10.14.6
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 2.26 GB / 32.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 12.12.0 - /var/folders/_9/nh0gc1bs08qd9p3kqshr3f0h0000gn/T/yarn--1581784553520-0.7678366513125479/node
    Yarn: 1.19.1 - /var/folders/_9/nh0gc1bs08qd9p3kqshr3f0h0000gn/T/yarn--1581784553520-0.7678366513125479/yarn
    npm: 6.11.3 - ~/.nvm/versions/node/v12.12.0/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 13.2, DriverKit 19.0, macOS 10.15, tvOS 13.2, watchOS 6.1
  IDEs:
    Android Studio: 3.5 AI-191.8026.42.35.5791312
    Xcode: 11.2.1/11B500 - /usr/bin/xcodebuild
  npmPackages:
    react: 16.9.0 => 16.9.0 
    react-native: 0.61.5 => 0.61.5
```

###  Steps To Reproduce

1. Create a `<FlatList />` with `<Text />` items in it
2. Make sure the list is longer than the screen
3. Check the first render of this list, scroll

###  Expected Results

Scrollbar should stick to the right of the screen.

###  Snack, code example, screenshot, or link to a repository:

Screenshots can be found in this issue.

**Comment by safaiyeh** [comment]
Hi @jasperkuperus can you open a new issue with the template you just provided and reference this one.

We have a PR from @billxie1988 https://github.com/react-native-community/react-native-webview/pull/1077 in `react-native-webview` that might be more appropriate to add to ScrollView instead

**Comment by jasperkuperus** [comment]
Sure, done. I'm curious though. Why can't you guys not simply re-open an issue? Opening a new issue and keeping this one closed only introduces more clutter and possible confusion for people searching for these issues, I'd say. @hramos @safaiyeh

**Comment by Elijen** [comment]
@jasperkuperus Gotta hit the KPIs!

**Comment by Rajibshrestha59** [comment]
> Is it any related issue tracking this? I don't think that it is a out of date problem :/
> 
> Device: iPhone 11 Pro Max with iOS 13.3/Simulator
> Version: ReactNative@0.61.5, React@16.9.0
> 
> ![untitled](https://user-images.githubusercontent.com/11359892/73992643-b4736f80-498a-11ea-956a-bb68b648d4ef.png) 

Please add scrollIndicatorInsets={{ right: 0 }} inside ScrollView tag
<ScrollView scrollIndicatorInsets={{ right: 0 }} > like this

**Comment by outaTiME** [comment]
Same here guys, thanks @sbycrosz for save my day, i prefer the `-1` value for right (better alignment imho).

```
<NativeScrollView
  {...extra}
  scrollIndicatorInsets={{ right: -1 }}
  contentInsetAdjustmentBehavior="automatic"
>
  {children}
</NativeScrollView>
```

before fix:

<img width="656" alt="Screen Shot 2020-04-02 at 00 28 45" src="https://user-images.githubusercontent.com/140472/78209384-3d83c000-747d-11ea-961b-ef22128b9efc.png">

after @sbycrosz workaround:

<img width="656" alt="Screen Shot 2020-04-02 at 01 00 07" src="https://user-images.githubusercontent.com/140472/78209570-c864ba80-747d-11ea-8638-ce17fc34f86b.png">

@hramos any update with this issue?

**Comment by 349989153** [comment]
Same issue

**Comment by chesovnik** [comment]
I have experienced the same issue with WebView on IOS 13. 
The issue was resolved by replacing WebView's prop "style" with "containerStyle"
`containerStyle={{
            flex: 1,
          }}`

**Comment by danilo900** [comment]
> I have experienced the same issue with WebView on iOS 13.
> The issue was resolved by replacing WebView's prop "style" with "containerStyle"
> `containerStyle={{ flex: 1 }}`

I tried here with no results on webview =/

**Comment by moh3n9595** [comment]
Same issue

**Comment by danilo900** [comment]
Nothing here?

**Comment by jimishshah** [comment]
Containing the app in `SafeAreaView` fixed my problem.

**Comment by Daniel-Griffiths** [comment]
I can also confirm that adding `scrollIndicatorInsets={{ right: 1 }}` to the ScrollView fixed the issue for me :slightly_smiling_face:

**Comment by Donhv** [comment]
I have same issue with react-native-webview.

**Comment by Donhv** [comment]
> I have experienced the same issue with WebView on iOS 13.
> The issue was resolved by replacing WebView's prop "style" with "containerStyle"
> `containerStyle={{ flex: 1 }}`
I have same issue with react-native-webview and i tried your solution but it didn't work

**Comment by Donhv** [comment]
> I have experienced the same issue with WebView on iOS 13.
> The issue was resolved by replacing WebView's prop "style" with "containerStyle"
> `containerStyle={{ flex: 1 }}`

I have same issue with react-native-webview and i tried your solution but it didn't work

**Comment by danilo900** [comment]
Yea. A bad issue on Webview and no solution till now.

**Comment by Donhv** [comment]
> Yea. A bad issue on Webview and no solution till now.

it maybe work if you wrap webview in SafeAreaView. but my solution is hide vertical scroll :D

**Comment by dani-mp** [comment]
Hi, @hramos. Sorry, I'm about to be a bit critical here. I can't believe this issue has been closed like that. This is not a minor bug, and I would say it's super important for React Native itself, not just its scroll view: this is the kind of bug that a native iOS/Flutter dev posts on Twitter and the React Native community gets one week of public shaming. I hope you understand what I mean.

**Comment by RageBill** [comment]
This issue should not be closed without having a proper solution yet.

This is very frustrating to see a bug from almost one year ago to still haunt me till these days.

I tried the fixes given by @sbycrosz and @outaTiME , and neither of them worked in my case.

I am on `RN 0.63.2` right now.

**Comment by safaiyeh** [comment]
Since this is coming up again, I'll reopen.

We had a PR open in webview to be ported to core: https://github.com/react-native-community/react-native-webview/pull/1077

Could someone open a PR and test if this fix would work for RN Core?

**Comment by renielDev** [comment]
Same issue I've been experiencing and adding  `scrollIndicatorInsets={{ right: 1 }}` to the ScrollView fixed it.

**Comment by billxie1988** [comment]
> Since this is coming up again, I'll reopen.
> 
> We had a PR open in webview to be ported to core: [react-native-community/react-native-webview#1077](https://github.com/react-native-community/react-native-webview/pull/1077)
> 
> Could someone open a PR and test if this fix would work for RN Core?

updated

**Comment by kylerc** [comment]
Currently seeing this issue with WebViews on iOS14, looks like all the solutions suggested so far haven't resolved it for me.

**Comment by hengkx** [comment]
IOS 14 sectionlist same

**Comment by niallbroadley** [comment]
> IOS 14 sectionlist same

Yeah me too. This actually nukes my app completely, it's impossible to release something commercially with this kind of bug present.

**Comment by osartun** [comment]
Linking this PR here, because it fixes the issue: https://github.com/facebook/react-native/pull/29809

**Comment by RodolfoGS** [comment]
Mar 20, 2021... no fixed yet 🤘

**Comment by saskaak** [comment]
The workaround suggested by @sbycrosz moves the scrollbar 1px to left. To keep original position, you can use:

```jsx
scrollIndicatorInsets={{right: Number.MIN_VALUE}}
```

**Comment by danilo900** [comment]
> The workaround suggested by @sbycrosz moves the scrollbar 1px to left. To keep original position, you can use:
> 
> ```js
> scrollIndicatorInsets={{right: Number.MIN_VALUE}}
> ```

How we will add that property to Webview?

**Comment by saskaak** [comment]
> > The workaround suggested by @sbycrosz moves the scrollbar 1px to left. To keep original position, you can use:
> > ```js
> > scrollIndicatorInsets={{right: Number.MIN_VALUE}}
> > ```
> 
> How we will add that property to Webview?

Dunno, this issue is about ScrollView.

**Comment by cheese-git** [comment]
add style `flexGrow: 1` to parent component fixed this in my situation:

```tsx
<View style={{ flexGrow: 1 }}>
  <FlatList />
</View>
```

**Comment by mlazari** [comment]
It looks like this issue happens when the ScrollView (and probably other scrollable components too) is mounted in a screen while the navigation animation is running. For example when rendering the ScrollView only after receiving data from an API. The position of the scroll bar seems to be set depending on where the ScrollView is on the screen when it's mounted / rendered. Here is a repro that simulates waiting for data with a `setTimeout`: https://snack.expo.dev/@l3m3a3/scrollview-bug Changing the timeout value changes where the scroll bar appears.
https://www.dropbox.com/s/yli1ion0mjdus2i/scroll-view-bug.mov?dl=0
Setting `automaticallyAdjustsScrollIndicatorInsets={false}` (available in React Native 0.66) for the `ScrollView` seems to fix the issue.

**Comment by mlazari** [comment]
It looks like this issue happens when the ScrollView (and probably other scrollable components too) is mounted in a screen while the navigation animation is running. For example when rendering the ScrollView only after receiving data from an API. The position of the scroll bar seems to be set depending on where the ScrollView is on the screen when it's mounted / rendered. Here is a repro that simulates waiting for data with a `setTimeout`: https://snack.expo.dev/@l3m3a3/scrollview-bug Changing the timeout value changes where the scroll bar appears.
https://www.dropbox.com/s/yli1ion0mjdus2i/scroll-view-bug.mov?dl=0
Setting `automaticallyAdjustsScrollIndicatorInsets={false}` (available in React Native 0.66) for the `ScrollView` seems to fix the issue, so it looks like it's related to this prop.

**Comment by mlazari** [comment]
So my fix for now will be to change the default value for `ScrollView`'s `automaticallyAdjustsScrollIndicatorInsets` prop to false at the beginning of the root index.js file. I tested it and this fixes the issue for `ScrollView`, `FlatList`, `SectionList`.

```
import { ScrollView } from 'react-native';

ScrollView.defaultProps = ScrollView.defaultProps || {};
ScrollView.defaultProps.automaticallyAdjustsScrollIndicatorInsets = false;
```

**Comment by rim-sila** [comment]
> Same here guys, thanks @sbycrosz for save my day, to preserve the pixel proportions, it would be better to use `top`/ `bottom` instead of `right`, here my code:
> 
> ```js
> <ScrollView
>   scrollIndicatorInsets={{ top: 1, bottom: 1 }}
>   contentInsetAdjustmentBehavior="always"
>   {...props}
> />
> ```
> 
> before fix:
> 
> <img alt="Screen Shot 2020-04-02 at 00 28 45" width="656" src="https://user-images.githubusercontent.com/140472/78209384-3d83c000-747d-11ea-961b-ef22128b9efc.png">
> 
> after @sbycrosz workaround:
> 
> <img alt="Screen Shot 2020-04-02 at 01 00 07" width="656" src="https://user-images.githubusercontent.com/140472/78209570-c864ba80-747d-11ea-8638-ce17fc34f86b.png">
> 
> @hramos any update with this issue?



> Same here guys, thanks @sbycrosz for save my day, to preserve the pixel proportions, it would be better to use `top`/ `bottom` instead of `right`, here my code:
> 
> ```js
> <ScrollView
>   scrollIndicatorInsets={{ top: 1, bottom: 1 }}
>   contentInsetAdjustmentBehavior="always"
>   {...props}
> />
> ```
> 
> before fix:
> 
> <img alt="Screen Shot 2020-04-02 at 00 28 45" width="656" src="https://user-images.githubusercontent.com/140472/78209384-3d83c000-747d-11ea-961b-ef22128b9efc.png">
> 
> after @sbycrosz workaround:
> 
> <img alt="Screen Shot 2020-04-02 at 01 00 07" width="656" src="https://user-images.githubusercontent.com/140472/78209570-c864ba80-747d-11ea-8638-ce17fc34f86b.png">
> 
> @hramos any update with this issue?

this working for me

**Comment by mvanroon** [comment]
This error occurred in our app when data was loaded while the react navigation screen transition was still underway, causing the screen to re-render mid-transition.

These solutions all fixed the problem:
 - load data before navigation
 - load data after transition
 - add `scrollIndicatorInsets={{ right: 1 }}` to the `ScrollView`
 - render the `ScrollView` with `flex:1` even when data hasn't finished loading yet

We went with the 4th solution.

**Comment by sn4f** [comment]
In my case adding 
```
// @ts-ignore
automaticallyAdjustsScrollIndicatorInsets={false}
```
to FlatList or ScrollView helped.

https://reactnative.dev/docs/scrollview#automaticallyadjustsscrollindicatorinsets-ios

**Comment by alexanderblackh** [comment]
This issue is still appearing, specifically in my case only when wrapped in a KeyboardAvoidingView. Configured as the following:

```
<KeyboardAvoidingView
      style={{flex: 1, flexDirection: 'column', justifyContent: 'center'}}
      behavior="padding"
      enabled
      keyboardVerticalOffset={100}>
    <ScrollView
        contentContainerStyle={{paddingHorizontal: spacing.m}}>
         {....}
    </ScrollView>
</KeyboardAvoidingView>
```

Can confirm @sbycrosz's solution of `scrollIndicatorInsets={{ right: 1 }}` did fix it.

**Comment by zholmes1** [comment]
For my instance of this bug, it was caused by conditionally rendering a FlatList. My pseudo-code was:

```
if (initialLoading) return <FullScreenLoadingPage/>
    
else return <FlatList data={...} />
```

When I re-worked this to:

return (
  <FlatList 
    data={...} 
    ListEmptyState={initialLoading ? FullScreenLoadingPage : ...} 
  />
)

The problem went away. I encourage you to try this solution before you start messing around with manually forcing the indicator insets.
